thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,141 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef TRACESLAVE_HPP
+#define TRACESLAVE_HPP
+
+#include "slave.hpp"
+#include "jqueue.tpp"
+#include "jthread.hpp"
+
+#define TenMinutesInMs 1000*60*10
+
+
+class RowRingBuffer: public SafeRingBufferOf<const void>
+{
+    void releaseBuffer(const void *e)
+    {
+        ReleaseThorRow(e);
+    };
+};
+
+class RowTracerThread : extends Thread
+{
+    RowRingBuffer *rowRingBuffer;
+    IHThorArg *helper;
+    Semaphore waitSemaphore;
+    bool runThread;
+    unsigned intervalInMSecs;
+    int activityId;
+    
+public:
+    RowTracerThread(IHThorArg *_helper,RowRingBuffer *_rowRingBuffer, unsigned _intervalInMSecs, unsigned _activityId)
+    : Thread("RowTracerThread"), helper(_helper), rowRingBuffer(_rowRingBuffer), 
+      intervalInMSecs(_intervalInMSecs), activityId(_activityId) {};
+
+    virtual int run()
+    {
+        runThread = true;
+        while (runThread)
+        {
+            waitSemaphore.wait(intervalInMSecs);
+            const void *row;
+
+            for (int maxRows=10; maxRows; --maxRows)
+            {
+                row = rowRingBuffer->dequeue();
```

Comment:
You should probably extract all rows from the buffer in a single atomic operation, otherwise you may get a list of rows that isn't actually contiguous in the datastream.   (Converting to xml and tracing could easily be slower than appending rows).